### ISSUE TITLE
CS0220 on uint.MaxValue + (nuint)42

### ISSUE THREAD

**Issue by lostmsu** [dotnet/roslyn#51714]
**Version Used**: 3.10.0-1.21125.6 (93b8af2b)

**Steps to Reproduce**: `uint.MaxValue + (nuint)42`

**Expected Behavior**:
`uint + nuint` => promote `uint` to `nuint`, then add.

**Actual Behavior**:  error CS0220: The operation overflows at compile time in checked mode

**Workaround**:
```csharp
uint uintMax = uint.MaxValue;
nuint size = uintMax + (nuint)42;
```

**Comment by Youssef1313** [comment]
> **Expected Behavior:**
> `uint + nuint` => promote `uint` to `nuint`, then add.

- There is no implicit conversion between uint to nuint per spec. Only explicit conversion exists.
- I can't see how converting uint to nuint would prevent the overflow from happening.
- Your workaround doesn't prevent the overflow from happening. The overflow would happen, but the compiler doesn't report it because `uintMax` isn't a compile-time constant, and there is no kind of data flow analysis done here.

**Comment by cston** [comment]
Thanks for reporting @lostmsu.

An error is reported for `uint.MaxValue + (nuint)42` because the expression overflows when `nuint` is 32 bits.

As @Youssef1313 described, the second example avoids the compile time error, but will still overflow at runtime when executed on a 32-bit platform.

Closing as "by design".

**Comment by lostmsu** [comment]
@cston this would make sense, if `nuint` was always 32 bit.

In fact, this directly contradicts the spec: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/native-integers#constants

>  Constant folding operations are evaluated with `Int32` and `UInt32` operands rather than native ints for consistent behavior regardless of compiler platform. If the operation results in a constant value in 32-bits, constant folding is performed at compile-time. Otherwise the operation is executed at runtime and not considered a constant.

**Comment by cston** [comment]
@lostmsu,

> this would make sense, if `nuint` was always 32 bit, and I was trying to define a `const nuint` field or variable. But I am just trying to pass that value to a function, that expects `nuint`.

Good point, thanks. 

From the [LDM notes](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-03-25.md#behavior-of-constant-folding):

> Whenever we can safely produce a constant value under 32-bits, we do constant folding. Otherwise, the result is non-constant, and under `checked`, the code produces a warning and the result is non-constant.

I'll investigate.
